server/src/main/java/org/eclipse/lsp/cobol/core/model/tree/variables/VariableDefinitionUtil.java

```diff
@@ -551,16 +559,38 @@ private VariableWithLevelNode getVariableForConditional(VariableDefinitionNode d
     return new ResultWithErrors<>(variableNode, errors);
   }
 
-  private VariableWithLevelNode getPrecedingVariableForRedefine(
-      VariableDefinitionNode definitionNode, String redefineName) {
+  private static List<VariableWithLevelNode> getEligibleNodesForRedefine(
+      VariableDefinitionNode definitionNode) {
     Deque<Node> siblings = new LinkedList<>(definitionNode.getParent().getChildren());
-    while (!siblings.isEmpty()) {
-      Node precedingNode = siblings.pollLast();
-      if (precedingNode instanceof VariableWithLevelNode) {
-        VariableWithLevelNode variable = (VariableWithLevelNode) precedingNode;
-        if (!variable.isRedefines() || variable.getName().equals(redefineName)) return variable;
-      }
+    List<VariableWithLevelNode> eligibleNodesForRedefine = new ArrayList<>();
+    LinkedList<VariableWithLevelNode> availableNodes =
+        siblings.stream()
+            .filter(node -> node instanceof VariableWithLevelNode)
+            .map(VariableWithLevelNode.class::cast)
+            .filter(
+                node ->
+                    !Objects.equals(node.getClass(), ConditionDataNameNode.class)
+                        && !Objects.equals(node.getClass(), RenameItemNode.class))
```

Comment:
Optionally, you may check this by comparing the `VariableType`

Reply:
updated